File: apis/elysium-character-api/src/main/java/com/seventh_root/elysium/api/character/CharacterProvider.java
package com.seventh_root.elysium.api.character;

import com.seventh_root.elysium.api.player.ElysiumPlayer;
import com.seventh_root.elysium.core.service.ServiceProvider;

import java.util.Collection;

public interface CharacterProvider<T extends ElysiumCharacter> extends ServiceProvider {

    T getCharacter(int id);

    T getActiveCharacter(ElysiumPlayer player);

    void setActiveCharacter(ElysiumPlayer player, T character);

    Collection<? extends T> getCharacters(ElysiumPlayer player);

    int addCharacter(T character);

    void removeCharacter(T character);

    void updateCharacter(T character);

}

File: apis/elysium-character-api/src/main/kotlin/com/seventh_root/elysium/api/character/CharacterProvider.kt
package com.seventh_root.elysium.api.character

import com.seventh_root.elysium.api.player.ElysiumPlayer
import com.seventh_root.elysium.core.service.ServiceProvider

interface CharacterProvider<T : ElysiumCharacter> : ServiceProvider {

    fun getCharacter(id: Int): T?
    fun getActiveCharacter(player: ElysiumPlayer): T?
    fun setActiveCharacter(player: ElysiumPlayer, character: T?)
    fun getCharacters(player: ElysiumPlayer): Collection<T>
    fun addCharacter(character: T): Int
    fun removeCharacter(character: T)
    fun updateCharacter(character: T)

}
